File: core/src/main/kotlin/com/backend/core/user/UserEntity.kt
package com.backend.core.user

import com.fasterxml.jackson.annotation.JsonIgnore
import jakarta.persistence.*
import org.hibernate.annotations.CreationTimestamp
import org.hibernate.annotations.UpdateTimestamp
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.UserDetails
import java.time.LocalDateTime
import java.util.*

@Entity
@Table(name = "users")
data class UserEntity(
    @Id
    @Column(columnDefinition = "UUID")
    val id: UUID = UUID.randomUUID(),

    @Column(unique = true, nullable = false)
    val email: String,

    @Column(unique = true)
    val username: String? = null,

    @JsonIgnore
    @Column(name = "hashed_password", nullable = false)
    val hashedPassword: String,

    @Column(name = "first_name")
    val firstName: String? = null,

    @Column(name = "last_name")
    val lastName: String? = null,

    // Profile fields
    @Column(name = "avatar_url")
    val avatarUrl: String? = null,

    val phone: String? = null,
    val title: String? = null,
    val department: String? = null,

    @Column(columnDefinition = "TEXT")
    val bio: String? = null,

    // Settings
    val timezone: String = "UTC",
    val locale: String = "en",
    val theme: String = "light",

    @Column(name = "email_notifications")
    val emailNotifications: Boolean = true,

    @Column(name = "sms_notifications")
    val smsNotifications: Boolean = false,

    @Column(name = "push_notifications")
    val pushNotifications: Boolean = true,

    @Column(name = "marketing_notifications")
    val marketingNotifications: Boolean = false,

    // Status fields
    @Column(name = "is_active")
    val isActive: Boolean = true,

    @Column(name = "is_verified")
    val isVerified: Boolean = false,

    @Column(name = "is_superuser")
    val isSuperuser: Boolean = false,

    // 2FA fields
    @Column(name = "two_factor_enabled")
    val twoFactorEnabled: Boolean = false,

    @JsonIgnore
    @Column(name = "two_factor_secret")
    val twoFactorSecret: String? = null,

    @JsonIgnore
    @Column(name = "backup_codes", columnDefinition = "TEXT")
    val backupCodes: String? = null, // JSON array of backup codes

    @Column(name = "two_factor_verified_at")
    val twoFactorVerifiedAt: LocalDateTime? = null,

    // Security fields
    @JsonIgnore
    @Column(name = "failed_login_attempts")
    val failedLoginAttempts: Int = 0,

    @JsonIgnore
    @Column(name = "locked_until")
    val lockedUntil: LocalDateTime? = null,

    @Column(name = "last_login")
    val lastLogin: LocalDateTime? = null,

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    val createdAt: LocalDateTime = LocalDateTime.now(),

    @UpdateTimestamp
    @Column(name = "updated_at")
    val updatedAt: LocalDateTime = LocalDateTime.now()
) : UserDetails {

    val fullName: String
        get() = "${firstName ?: ""} ${lastName ?: ""}".trim()

    @JsonIgnore
    override fun getAuthorities(): MutableCollection<out GrantedAuthority> {
        val authorities = mutableListOf<GrantedAuthority>()
        authorities.add(SimpleGrantedAuthority("ROLE_USER"))
        if (isSuperuser) {
            authorities.add(SimpleGrantedAuthority("ROLE_ADMIN"))
        }
        return authorities
    }

    @JsonIgnore
    override fun getPassword(): String = hashedPassword

    @JsonIgnore
    override fun getUsername(): String = email

    @JsonIgnore
    override fun isAccountNonExpired(): Boolean = true

    @JsonIgnore
    override fun isAccountNonLocked(): Boolean {
        return lockedUntil?.isBefore(LocalDateTime.now()) != false
    }

    @JsonIgnore
    override fun isCredentialsNonExpired(): Boolean = true

    @JsonIgnore
    override fun isEnabled(): Boolean = isActive && isVerified
}

File: identity-service/src/user/entity/user.entity.ts
import { Exclude } from 'class-transformer';
import {
  Column,
  CreateDateColumn,
  Entity,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
} from 'typeorm';

@Entity('users')
export class User {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column({ unique: true, nullable: false })
  email: string;

  @Column({ unique: true, nullable: true })
  username?: string;

  @Exclude()
  @Column({ name: 'hashed_password', nullable: false })
  hashedPassword: string;

  @Column({ name: 'first_name', nullable: true })
  firstName?: string;

  @Column({ name: 'last_name', nullable: true })
  lastName?: string;

  // Profile fields
  @Column({ name: 'avatar_url', nullable: true })
  avatarUrl?: string;

  @Column({ nullable: true })
  phone?: string;

  @Column({ nullable: true })
  title?: string;

  @Column({ nullable: true })
  department?: string;

  @Column({ type: 'text', nullable: true })
  bio?: string;

  // Settings
  @Column({ default: 'UTC' })
  timezone: string;

  @Column({ default: 'en' })
  locale: string;

  @Column({ default: 'light' })
  theme: string;

  @Column({ name: 'email_notifications', default: true })
  emailNotifications: boolean;

  @Column({ name: 'sms_notifications', default: false })
  smsNotifications: boolean;

  @Column({ name: 'push_notifications', default: true })
  pushNotifications: boolean;

  @Column({ name: 'marketing_notifications', default: false })
  marketingNotifications: boolean;

  // Status fields
  @Column({ name: 'is_active', default: true })
  isActive: boolean;

  @Column({ name: 'is_verified', default: false })
  isVerified: boolean;

  @Column({ name: 'is_superuser', default: false })
  isSuperuser: boolean;

  // Two-Factor Authentication
  @Column({ name: 'two_factor_enabled', default: false })
  twoFactorEnabled: boolean;

  @Exclude()
  @Column({ name: 'two_factor_secret', nullable: true })
  twoFactorSecret?: string;

  @Exclude()
  @Column({ name: 'backup_codes', type: 'text', nullable: true })
  backupCodes?: string; // JSON array of backup codes

  @Column({ name: 'two_factor_verified_at', type: 'timestamp', nullable: true })
  twoFactorVerifiedAt?: Date;

  // Security fields
  @Exclude()
  @Column({ name: 'failed_login_attempts', default: 0 })
  failedLoginAttempts: number;

  @Exclude()
  @Column({ name: 'locked_until', type: 'timestamp', nullable: true })
  lockedUntil?: Date;

  @Column({ name: 'last_login', type: 'timestamp', nullable: true })
  lastLogin?: Date;

  // Auth0 Integration
  @Column({ name: 'auth0_id', nullable: true, unique: true })
  auth0Id?: string;

  // Timestamps
  @CreateDateColumn({ name: 'created_at' })
  createdAt: Date;

  @UpdateDateColumn({ name: 'updated_at' })
  updatedAt: Date;

  // Virtual properties (similar to Kotlin)
  get fullName(): string {
    return `${this.firstName || ''} ${this.lastName || ''}`.trim();
  }

  get isAccountNonLocked(): boolean {
    if (!this.lockedUntil) return true;
    return new Date() > this.lockedUntil;
  }

  get isEnabled(): boolean {
    return this.isActive && this.isVerified;
  }

  get authorities(): string[] {
    const authorities = ['ROLE_USER'];
    if (this.isSuperuser) {
      authorities.push('ROLE_ADMIN');
    }
    return authorities;
  }
}
